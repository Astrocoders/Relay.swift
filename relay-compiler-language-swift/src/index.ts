import { PluginInterface } from 'relay-compiler/lib/language/RelayLanguagePluginInterface';
import { find } from './FindGraphQLTags';
import { SwiftGenerator } from './SwiftGenerator';
import { formatGeneratedModule } from './formatGeneratedModule';
import { RunState } from './RunState';

import * as path from 'path';
import * as fs from 'fs';

export default function plugin(): PluginInterface {
  const runState = new RunState();
  return {
    inputExtensions: ['swift'],
    outputExtension: 'swift',
    getFileFilter(baseDir: string) {
      return (file: any) => {
        runState.reset();
        const filePath = path.join(baseDir, file.relPath);
        let text = '';
        try {
          text = fs.readFileSync(filePath, 'utf8');
        } catch {
          // eslint-disable no-console
          console.warn(
            `RelaySourceModuleParser: Unable to read the file "${filePath}". Looks like it was removed.`
          );
          return false;
        }
        return text.indexOf('graphql') >= 0;
      };
    },
    findGraphQLTags: find,
    formatModule: formatGeneratedModule(),
    typeGenerator: SwiftGenerator(runState),
  } as any;
}
