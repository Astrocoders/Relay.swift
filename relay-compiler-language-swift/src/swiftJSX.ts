export interface SwiftElement {
  type: string | SwiftComponent;
  props: any;
}
export type SwiftChild = SwiftElement | string;
export type SwiftNode = null | undefined | SwiftChild | SwiftFragment;
export type SwiftFragment = SwiftNode[];
export type SwiftComponent<Props = any> = (props: Props) => SwiftElement | null;

export function swiftJSX(
  tagName: string | SwiftComponent,
  attrs: any,
  ...children: SwiftNode[]
): SwiftElement | null {
  return { type: tagName, props: { ...attrs, children } };
}

export const Fragment = 'swiftFragment';

export function renderSwift(element: SwiftElement): string {
  const renderer = new Renderer();
  _renderSwift(renderer, element);
  return renderer.text;
}

interface RenderContext {
  inline?: boolean;
  inProtocol?: boolean;
  inSwitch?: boolean;
}

function _renderSwift(
  renderer: Renderer,
  element: SwiftNode,
  context: RenderContext = {}
) {
  if (!element) {
    return;
  }

  if (Array.isArray(element)) {
    for (const node of element) {
      _renderSwift(renderer, node, context);
    }
    return;
  }

  if (typeof element === 'string') {
    if (context.inline) {
      renderer.append(element);
    } else {
      renderer.appendLine(element);
    }
    return;
  }

  if (typeof element.type === 'string') {
    switch (element.type) {
      case 'swiftFragment':
        return _renderSwift(renderer, element.props.children);
      case 'init':
        return renderSwiftInit(renderer, element.props);
      case 'function':
        return renderSwiftFunction(renderer, element.props);
      case 'paramdecl':
        return renderSwiftFunctionParam(renderer, element.props);
      case 'struct':
        return renderSwiftStruct(renderer, element.props);
      case 'var':
        return renderSwiftVar(renderer, element.props, context);
      case 'extension':
        return renderSwiftExtension(renderer, element.props);
      case 'protocol':
        return renderSwiftProtocol(renderer, element.props);
      case 'enum':
        return renderSwiftEnum(renderer, element.props);
      case 'case':
        return renderSwiftCase(renderer, element.props, context);
      case 'import':
        return renderSwiftImport(renderer, element.props);
      case 'available':
        return renderSwiftAvailable(renderer, element.props);
      case 'call':
        return renderSwiftCall(renderer, element.props);
      case 'param':
        return renderSwiftParam(renderer, element.props, context);
      case 'compilecheck':
        return renderSwiftCompileTimeCheck(renderer, element.props);
      case 'typealias':
        return renderSwiftTypealias(renderer, element.props);
      case 'literal':
        return renderSwiftLiteral(renderer, element.props, context);
      case 'switch':
        return renderSwiftSwitch(renderer, element.props);
    }
  } else {
    const rendered = element.type(element.props);
    if (!rendered) {
      return;
    }
    return _renderSwift(renderer, rendered, context);
  }
}

class Renderer {
  private lines: string[];
  private level: number;
  private shouldContinue: boolean;

  constructor() {
    this.lines = [];
    this.level = 0;
    this.shouldContinue = false;
  }

  appendLine(line: string) {
    if (this.shouldContinue) {
      this.append(line);
    } else {
      this.lines.push('    '.repeat(this.level) + line);
    }
  }

  append(text: string) {
    if (this.shouldContinue) {
      this.shouldContinue = false;
    }

    if (this.lines.length) {
      this.lines[this.lines.length - 1] += text;
    } else {
      this.appendLine(text);
    }
  }

  continueLine() {
    this.shouldContinue = true;
  }

  indent() {
    this.level++;
  }

  outdent() {
    this.level--;
  }

  get text(): string {
    return this.lines.join('\n');
  }
}

type AccessLevel = 'public' | 'internal' | 'fileprivate' | 'private';

function renderSwiftStruct(
  renderer: Renderer,
  {
    name,
    inherit = [],
    access,
    children,
  }: {
    name: string;
    inherit?: string[];
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(
    `${access ? `${access} ` : ''}struct ${name}${
      inherit.length ? `: ${inherit.join(', ')}` : ''
    } {`
  );
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftVar(
  renderer: Renderer,
  {
    name,
    type,
    isStatic = false,
    access,
    children,
  }: {
    name: string;
    type: string;
    isStatic?: boolean;
    access?: AccessLevel;
    children: SwiftNode[];
  },
  context: RenderContext = {}
) {
  renderer.appendLine(
    `${access ? `${access} ` : ''}${
      isStatic ? 'static ' : ''
    }var ${name}: ${type}`
  );

  if (context.inProtocol) {
    renderer.append(' { get }');
    return;
  }

  if (children.length) {
    renderer.append(' {');
    renderer.indent();

    _renderSwift(renderer, children, context);

    renderer.outdent();
    renderer.appendLine('}');
  }
}

function renderSwiftFunction(
  renderer: Renderer,
  {
    name,
    parameters = [],
    returns,
    isStatic = false,
    access,
    children,
  }: {
    name: string;
    parameters: SwiftNode[];
    returns?: string;
    isStatic?: boolean;
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(
    `${access ? `${access} ` : ''}${isStatic ? 'static ' : ''}func ${name}(`
  );
  for (let i = 0; i < parameters.length; i++) {
    if (i !== 0) {
      renderer.append(', ');
    }
    _renderSwift(renderer, parameters[i], { inline: true });
  }
  renderer.append(`)`);

  if (returns) {
    renderer.append(` -> ${returns}`);
  }

  renderer.append(` {`);
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftInit(
  renderer: Renderer,
  {
    parameters = [],
    access,
    throws = false,
    children,
  }: {
    parameters: SwiftNode[];
    access?: AccessLevel;
    throws: boolean;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}init(`);
  for (let i = 0; i < parameters.length; i++) {
    if (i !== 0) {
      renderer.append(', ');
    }
    _renderSwift(renderer, parameters[i], { inline: true });
  }
  renderer.append(')');
  if (throws) {
    renderer.append(' throws');
  }
  renderer.append(' {');
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftFunctionParam(
  renderer: Renderer,
  {
    label,
    name,
    type,
    defaultValue,
  }: {
    label?: string;
    name: string;
    type: string;
    defaultValue?: SwiftNode;
  }
) {
  if (label) {
    renderer.append(`${label} `);
  }
  renderer.append(`${name}: ${type}`);
  if (defaultValue) {
    renderer.append(' = ');
    _renderSwift(renderer, defaultValue, { inline: true });
  }
}

function renderSwiftExtension(
  renderer: Renderer,
  {
    name,
    inherit = [],
    where = [],
    access,
    children,
  }: {
    name: string;
    inherit?: string[];
    where?: string[];
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}extension ${name}`);
  if (inherit.length) {
    renderer.append(`: ${inherit.join(', ')}`);
  }
  if (where.length) {
    renderer.append(` where ${where.join(', ')}`);
  }
  renderer.append(' {');

  if (!children.length) {
    renderer.append('}');
    return;
  }

  renderer.indent();
  _renderSwift(renderer, children);
  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftProtocol(
  renderer: Renderer,
  {
    name,
    access,
    children,
  }: {
    name: string;
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}protocol ${name} {`);
  renderer.indent();

  _renderSwift(renderer, children, { inProtocol: true });

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftEnum(
  renderer: Renderer,
  {
    name,
    inherit = [],
    access,
    children,
  }: {
    name: string;
    inherit: string[];
    access?: AccessLevel;
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}enum ${name}`);
  if (inherit.length) {
    renderer.append(`: ${inherit.join(', ')}`);
  }
  renderer.append(' {');
  renderer.indent();

  _renderSwift(renderer, children);

  renderer.outdent();
  renderer.appendLine('}');
}

function renderSwiftCase(
  renderer: Renderer,
  {
    name,
    parameters = [],
    isDefault = false,
    children,
  }: {
    name?: SwiftNode;
    parameters?: SwiftNode[];
    isDefault: boolean;
    children: SwiftNode[];
  },
  context: RenderContext = {}
) {
  if (isDefault) {
    renderer.appendLine('default');
  } else {
    renderer.appendLine('case ');
    _renderSwift(renderer, name, { inline: true });

    if (parameters.length) {
      renderer.append('(');

      for (let i = 0; i < parameters.length; i++) {
        if (i !== 0) {
          renderer.append(', ');
        }
        _renderSwift(renderer, parameters[i], { inline: true });
      }

      renderer.append(')');
    }
  }

  if (context.inSwitch) {
    renderer.append(':');
    renderer.indent();

    _renderSwift(renderer, children, context);

    renderer.outdent();
  } else {
    if (children.length) {
      renderer.append(' = ');
      _renderSwift(renderer, children, { inline: true });
    }
  }
}

function renderSwiftImport(
  renderer: Renderer,
  {
    module,
  }: {
    module: string;
  }
) {
  renderer.appendLine(`import ${module}`);
}

function renderSwiftAvailable(
  renderer: Renderer,
  {
    versions,
    children,
  }: {
    versions: string[];
    children: SwiftNode[];
  }
) {
  renderer.appendLine(`@available(${versions.join(', ')}, *)`);
  _renderSwift(renderer, children);
}

function renderSwiftCall(
  renderer: Renderer,
  {
    receiver,
    name,
    parameters = [],
    expanded = false,
  }: {
    receiver?: SwiftNode;
    name: string;
    parameters?: SwiftNode[];
    expanded?: boolean;
  }
) {
  renderer.appendLine(`${receiver != null ? `${receiver}.` : ''}${name}(`);
  if (expanded) {
    renderer.indent();
  }

  parameters = parameters.filter(p => !!p);
  for (let i = 0; i < parameters.length; i++) {
    if (i !== 0) {
      renderer.append(',');
      if (!expanded) {
        renderer.append(' ');
      }
    }
    _renderSwift(renderer, parameters[i], { inline: !expanded });
  }

  if (expanded) {
    renderer.outdent();
    renderer.appendLine(')');
  } else {
    renderer.append(')');
  }
}

function renderSwiftParam(
  renderer: Renderer,
  {
    label,
    children,
  }: {
    label?: string;
    children: SwiftNode[];
  },
  context: RenderContext = {}
) {
  const text = `${label ? `${label}: ` : ''}`;
  if (context.inline) {
    renderer.append(text);
  } else {
    renderer.appendLine(text);
  }

  renderer.continueLine();
  _renderSwift(renderer, children, { ...context, inline: true });
}

function renderSwiftCompileTimeCheck(
  renderer: Renderer,
  {
    condition,
    children,
  }: {
    condition: SwiftNode;
    children: SwiftNode[];
  }
) {
  renderer.appendLine('#if ');

  _renderSwift(renderer, condition, { inline: true });
  _renderSwift(renderer, children);

  renderer.appendLine('#endif');
}

function renderSwiftTypealias(
  renderer: Renderer,
  {
    name,
    children,
    access,
  }: {
    name: string;
    children: SwiftNode[];
    access?: AccessLevel;
  }
) {
  renderer.appendLine(`${access ? `${access} ` : ''}typealias ${name} = `);
  _renderSwift(renderer, children, { inline: true });
}

type SwiftLiteralProps =
  | { string: string }
  | { int: number }
  | { bool: boolean }
  | { array: SwiftNode[]; expanded?: boolean }
  | { dict: [SwiftNode, SwiftNode][]; expanded?: boolean };

function renderSwiftLiteral(
  renderer: Renderer,
  props: SwiftLiteralProps,
  context: RenderContext
) {
  const appendLine = context.inline
    ? text => {
        renderer.append(text);
      }
    : text => {
        renderer.appendLine(text);
      };

  if ('string' in props) {
    appendLine(JSON.stringify(props.string));
  } else if ('int' in props) {
    appendLine(String(props.int));
  } else if ('bool' in props) {
    appendLine(String(props.bool));
  } else if ('array' in props) {
    appendLine('[');
    renderer.indent();

    const innerContext = { ...context, inline: !props.expanded };

    for (let i = 0; i < props.array.length; i++) {
      if (i !== 0) {
        renderer.append(',');
        if (innerContext.inline) {
          renderer.append(' ');
        }
      }

      _renderSwift(renderer, props.array[i], innerContext);
    }

    renderer.outdent();
    if (props.expanded) {
      renderer.appendLine(']');
    } else {
      renderer.append(']');
    }
  } else if ('dict' in props) {
    appendLine('[');
    renderer.indent();

    const innerContext = { ...context, inline: !props.expanded };

    for (let i = 0; i < props.dict.length; i++) {
      if (i !== 0) {
        renderer.append(',');
        if (innerContext.inline) {
          renderer.append(' ');
        }
      }

      const [key, value] = props.dict[i];
      _renderSwift(renderer, key, innerContext);
      renderer.append(': ');
      renderer.continueLine();
      _renderSwift(renderer, value, innerContext);
    }

    renderer.outdent();
    if (props.expanded) {
      renderer.appendLine(']');
    } else {
      renderer.append(']');
    }
  } else {
    throw new Error('unrecognized props for literal');
  }
}

function renderSwiftSwitch(
  renderer: Renderer,
  {
    value,
    children,
  }: {
    value: SwiftNode;
    children: SwiftNode[];
  }
) {
  renderer.appendLine('switch ');
  _renderSwift(renderer, value, { inline: true });
  renderer.append(' {');

  _renderSwift(renderer, children, { inSwitch: true });

  renderer.appendLine('}');
}
