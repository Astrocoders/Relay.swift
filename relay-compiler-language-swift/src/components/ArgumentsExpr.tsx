/** @jsx swiftJSX */

import { swiftJSX } from '../swiftJSX';
import { ReaderArgument, NormalizationArgument } from 'relay-runtime';
import { ValueLiteral } from './ValueLiteral';

export const ArgumentsExpr = ({
  args,
}: {
  args: readonly ReaderArgument[] | readonly NormalizationArgument[];
}) => {
  return (
    <literal
      expanded
      array={(args as any).map(arg => {
        switch (arg.kind) {
          case 'Literal':
            return (
              <call
                name="LiteralArgument"
                parameters={[
                  <param label="name">
                    <literal string={arg.name} />
                  </param>,
                  <param label="value">
                    <ValueLiteral value={arg.value} />
                  </param>,
                ]}
              />
            );
          case 'Variable':
            return (
              <call
                name="VariableArgument"
                parameters={[
                  <param label="name">
                    <literal string={arg.name} />
                  </param>,
                  <param label="variableName">
                    <literal string={arg.variableName} />
                  </param>,
                ]}
              />
            );
          case 'ListValue':
            return (
              <call
                name="ListValueArgument"
                parameters={[
                  <param label="name">
                    <literal string={arg.name} />
                  </param>,
                  <param label="items">
                    <ArgumentsExpr args={arg.items} />
                  </param>,
                ]}
              />
            );
          case 'ObjectValue':
            return (
              <call
                name="ObjectValueArgument"
                parameters={[
                  <param label="name">
                    <literal string={arg.name} />
                  </param>,
                  <param label="fields">
                    <ArgumentsExpr args={arg.fields} />
                  </param>,
                ]}
              />
            );
        }
      })}
    />
  );
};
