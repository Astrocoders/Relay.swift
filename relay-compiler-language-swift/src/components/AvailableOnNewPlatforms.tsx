/** @jsx swiftJSX */

import { swiftJSX, SwiftNode } from '../swiftJSX';

export const AvailableOnNewPlatforms = ({
  children,
}: {
  children?: SwiftNode;
}) => {
  return (
    <available
      versions={['iOS 14.0', 'macOS 10.16', 'tvOS 14.0', 'watchOS 7.0']}
    >
      {children}
    </available>
  );
};
