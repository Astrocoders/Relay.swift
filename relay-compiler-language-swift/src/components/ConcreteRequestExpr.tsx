/** @jsx swiftJSX */

import { swiftJSX } from '../swiftJSX';
import { ConcreteRequest } from 'relay-runtime';
import { ReaderFragmentExpr } from './ReaderFragmentExpr';
import { NormalizationOperationExpr } from './NormalizationOperationExpr';
import { RequestParametersExpr } from './RequestParametersExpr';

export const ConcreteRequestExpr = ({ node }: { node: ConcreteRequest }) => {
  return (
    <call
      name="ConcreteRequest"
      parameters={[
        <param label="fragment">
          <ReaderFragmentExpr node={node.fragment} />
        </param>,
        <param label="operation">
          <NormalizationOperationExpr node={node.operation} />
        </param>,
        <param label="params">
          <RequestParametersExpr params={node.params} />
        </param>,
      ]}
      expanded
    />
  );
};
