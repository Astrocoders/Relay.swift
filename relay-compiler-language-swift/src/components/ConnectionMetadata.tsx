/** @jsx swiftJSX */

import { swiftJSX } from '../swiftJSX';
import { ReaderPaginationMetadata } from 'relay-runtime';
import { ConnectionVariableConfig } from './ConnectionVariableConfig';

export const ConnectionMetadata = ({
  metadata,
}: {
  metadata: ReaderPaginationMetadata;
}) => {
  return (
    <call
      name="ConnectionMetadata"
      parameters={[
        <param label="path">
          <literal
            array={metadata.path.map(s => (
              <literal string={s} />
            ))}
          />
        </param>,
        metadata.forward ? (
          <param label="forward">
            <ConnectionVariableConfig
              count={metadata.forward.count}
              cursor={metadata.forward.cursor}
            />
          </param>
        ) : null,
        metadata.backward ? (
          <param label="backward">
            <ConnectionVariableConfig
              count={metadata.backward.count}
              cursor={metadata.backward.cursor}
            />
          </param>
        ) : null,
      ]}
      expanded
    />
  );
};
