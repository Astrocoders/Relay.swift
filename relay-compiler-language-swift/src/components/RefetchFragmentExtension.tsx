/** @jsx swiftJSX */

import { swiftJSX } from '../swiftJSX';
import { ReaderFragment } from 'relay-runtime';
import { RefetchMetadata } from './RefetchMetadata';

export const RefetchFragmentExtension = ({
  node,
}: {
  node: ReaderFragment;
}) => {
  const refetch = node.metadata.refetch;
  const operationName = (refetch.operation as string)
    .replace('@@MODULE_START@@', '')
    .replace('.graphql@@MODULE_END@@', '');

  return (
    <extension
      name={node.name}
      inherit={[
        refetch.connection
          ? 'Relay.PaginationFragment'
          : 'Relay.RefetchFragment',
      ]}
    >
      <typealias name="Operation">{operationName}</typealias>
      <var isStatic name="metadata" type="Metadata">
        <RefetchMetadata metadata={refetch} />
      </var>
    </extension>
  );
};
