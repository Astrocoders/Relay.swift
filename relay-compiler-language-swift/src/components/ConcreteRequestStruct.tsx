/** @jsx swiftJSX */

import { swiftJSX } from '../swiftJSX';
import { ConcreteRequest } from 'relay-runtime';
import { ConcreteRequestExpr } from './ConcreteRequestExpr';

export const ConcreteRequestStruct = ({ node }: { node: ConcreteRequest }) => {
  return (
    <struct name={node.operation.name}>
      <var name="variables" type="Variables" />
      <init parameters={[<paramdecl name="variables" type="Variables" />]}>
        self.variables = variables
      </init>
      <var isStatic name="node" type="ConcreteRequest">
        <ConcreteRequestExpr node={node} />
      </var>
    </struct>
  );
};
