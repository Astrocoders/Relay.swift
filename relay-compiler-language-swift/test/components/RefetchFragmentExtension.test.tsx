/** @jsx swiftJSX */

import { swiftJSX, renderSwift } from '../../src/swiftJSX';
import { ReaderFragment } from 'relay-runtime';
import { RefetchFragmentExtension } from '../../src/components/RefetchFragmentExtension';

test('refetchable fragment', () => {
  const node: ReaderFragment = {
    kind: 'Fragment',
    name: 'TweetDetail_tweet',
    type: 'TweetGroup',
    metadata: {
      refetch: {
        fragmentPathInResult: ['node'],
        identifierField: 'id',
        operation:
          '@@MODULE_START@@TweetDetailRefetchQuery.graphql@@MODULE_END@@',
      } as any,
    },
    argumentDefinitions: [],
    selections: [],
  };

  const code = <RefetchFragmentExtension node={node} />;
  expect(renderSwift(code)).toMatchSnapshot();
});

test('pagination fragment', () => {
  const node: ReaderFragment = {
    kind: 'Fragment',
    name: 'TweetList_tweets',
    type: 'Viewer',
    metadata: {
      refetch: {
        fragmentPathInResult: ['viewer'],
        operation:
          '@@MODULE_START@@TweetListPaginationQuery.graphql@@MODULE_END@@',
        connection: {
          path: ['allTweets'],
          forward: { count: 'count', cursor: 'cursor' },
        },
      } as any,
    },
    argumentDefinitions: [],
    selections: [],
  };

  const code = <RefetchFragmentExtension node={node} />;
  expect(renderSwift(code)).toMatchSnapshot();
});
