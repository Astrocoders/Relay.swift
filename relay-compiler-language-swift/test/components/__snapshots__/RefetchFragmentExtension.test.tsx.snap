// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination fragment 1`] = `
"extension TweetList_tweets: Relay.PaginationFragment {
    typealias Operation = TweetListPaginationQuery
    static var metadata: Metadata {
        RefetchMetadata(
            path: [\\"viewer\\"],
            operation: Operation.self,
            connection: ConnectionMetadata(
                path: [\\"allTweets\\"],
                forward: ConnectionVariableConfig(count: \\"count\\", cursor: \\"cursor\\")
            )
        )
    }
}"
`;

exports[`refetchable fragment 1`] = `
"extension TweetDetail_tweet: Relay.RefetchFragment {
    typealias Operation = TweetDetailRefetchQuery
    static var metadata: Metadata {
        RefetchMetadata(
            path: [\\"node\\"],
            identifierField: \\"id\\",
            operation: Operation.self
        )
    }
}"
`;
